# **Final Assignment**
In this final assignment from IBM Professional Certificate Task by Adryan Rizky Alviano will comparing the result from machine learning model with the detail:



1**. Dataset**\
Dataset will be provided by IBM thats contain 40000 images with class of 2. The data will be split to train with 30000 images and test with 10000 images.

**2. The Task**\
There are 2 pre-trained model that will be use in this notebook.
*   Resnet50
*   VGG16

The Task will be:
*   Fit Train Model Using Resnet50
*   Fit Train Model Using VGG16
*   Fit Test Bot Model
*   Evaluate and Compare The Result

**3. Environment**\
Google colab will be used and setting hardware accelerator on GPU. 















## Library

In [ ]:
from tensorflow import keras
from keras.models import Sequential
from keras.layers import Dense
from keras import applications
from tensorflow.keras.applications import ResNet50
from keras.applications.resnet50 import preprocess_input
from keras.preprocessing.image import ImageDataGenerator

In [ ]:
import os

## **Import the Data**

In [ ]:
## get the data
!wget https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0321EN/data/concrete_data_week3.zip

--2021-06-28 00:20:35--  https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0321EN/data/concrete_data_week3.zip
Resolving s3-api.us-geo.objectstorage.softlayer.net (s3-api.us-geo.objectstorage.softlayer.net)... 67.228.254.196
Connecting to s3-api.us-geo.objectstorage.softlayer.net (s3-api.us-geo.objectstorage.softlayer.net)|67.228.254.196|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 261482368 (249M) [application/zip]
Saving to: ‘concrete_data_week3.zip’

concrete_data_week3 100%[===================>] 249.37M  24.3MB/s    in 10s     

2021-06-28 00:20:46 (24.0 MB/s) - ‘concrete_data_week3.zip’ saved [261482368/261482368]



In [ ]:
!unzip concrete_data_week3.zip

## **The Parameter**

In [ ]:
num_classes = 2

image_resize = 224

batch_size_training = 100
batch_size_validation = 100

## **Data Preprocessing**

In [ ]:
data_generator = ImageDataGenerator(
    preprocessing_function=preprocess_input,
)

In [ ]:
train_generator = data_generator.flow_from_directory(
    'concrete_data_week3/train',
    target_size=(image_resize, image_resize),
    batch_size=batch_size_training,
    class_mode='categorical')

Found 30001 images belonging to 2 classes.


In [ ]:
validation_generator = data_generator.flow_from_directory(
    'concrete_data_week3/valid',
    target_size=(image_resize, image_resize),
    batch_size=batch_size_training,
    class_mode='categorical')

Found 10001 images belonging to 2 classes.


# **Restnet50 Model**

In [ ]:
model = Sequential()

In [ ]:
model.add(ResNet50(
    include_top=False,
    pooling='avg',
    weights='imagenet',
    ))

94773248/94765736 [==============================] - 2s 0us/step


In [ ]:
model.add(Dense(num_classes, activation='softmax'))

In [ ]:
model.layers[0].trainable = False

In [ ]:
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

In [ ]:
steps_per_epoch_training = len(train_generator)
steps_per_epoch_validation = len(validation_generator)
num_epochs = 5

In [ ]:
fit_history = model.fit(
    train_generator,
    steps_per_epoch=steps_per_epoch_training,
    epochs=num_epochs,
    validation_data=validation_generator,
    validation_steps=steps_per_epoch_validation,
    verbose=1,
)

Epoch 1/5
301/301 [==============================] - 146s 484ms/step - loss: 0.0115 - accuracy: 0.9964 - val_loss: 0.0088 - val_accuracy: 0.9976
Epoch 2/5
301/301 [==============================] - 144s 479ms/step - loss: 0.0097 - accuracy: 0.9969 - val_loss: 0.0087 - val_accuracy: 0.9976
Epoch 3/5
301/301 [==============================] - 144s 479ms/step - loss: 0.0093 - accuracy: 0.9970 - val_loss: 0.0072 - val_accuracy: 0.9978
Epoch 4/5
301/301 [==============================] - 144s 479ms/step - loss: 0.0077 - accuracy: 0.9978 - val_loss: 0.0057 - val_accuracy: 0.9978
Epoch 5/5
301/301 [==============================] - 144s 480ms/step - loss: 0.0057 - accuracy: 0.9985 - val_loss: 0.0055 - val_accuracy: 0.9980


In [ ]:
model.save('classifier_resnet_model')

/usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/utils/generic_utils.py:497: CustomMaskWarning: Custom mask layers require a config and must override get_config. When loading, the custom mask layer must be passed to the custom_objects argument.
  category=CustomMaskWarning)


INFO:tensorflow:Assets written to: classifier_resnet_model/assets


# **VGG16 Model**

In [ ]:
from tensorflow.keras.applications import vgg16
from keras.applications.vgg16 import preprocess_input

In [ ]:
## get the data
!wget https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0321EN/data/concrete_data_week4.zip

--2021-06-28 00:32:30--  https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0321EN/data/concrete_data_week4.zip
Resolving s3-api.us-geo.objectstorage.softlayer.net (s3-api.us-geo.objectstorage.softlayer.net)... 67.228.254.196
Connecting to s3-api.us-geo.objectstorage.softlayer.net (s3-api.us-geo.objectstorage.softlayer.net)|67.228.254.196|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 261483817 (249M) [application/zip]
Saving to: ‘concrete_data_week4.zip’

concrete_data_week4 100%[===================>] 249.37M  19.7MB/s    in 12s     

2021-06-28 00:32:43 (20.2 MB/s) - ‘concrete_data_week4.zip’ saved [261483817/261483817]



In [ ]:
!unzip concrete_data_week4.zip

In [ ]:
data_generator2 = ImageDataGenerator(
    preprocessing_function=preprocess_input,
)

In [ ]:
train_generator2 = data_generator.flow_from_directory(
    'concrete_data_week4/train',
    target_size=(image_resize, image_resize),
    batch_size=batch_size_training,
    class_mode='categorical')

Found 30001 images belonging to 2 classes.


In [ ]:
validation_generator2 = data_generator.flow_from_directory(
    'concrete_data_week4/valid',
    target_size=(image_resize, image_resize),
    batch_size=batch_size_training,
    class_mode='categorical')

Found 9501 images belonging to 2 classes.


In [ ]:
model2 = Sequential()

In [ ]:
model2.add(vgg16.VGG16(
    include_top=False,
    pooling='avg',
    weights='imagenet'
    ))

58892288/58889256 [==============================] - 1s 0us/step


In [ ]:
model2.add(Dense(num_classes, activation='softmax'))

In [ ]:
model2.layers[0].trainable = False

In [ ]:
model2.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

In [ ]:
steps_per_epoch_training = len(train_generator)
steps_per_epoch_validation = len(validation_generator)
num_epochs = 5

In [ ]:
fit_history2 = model2.fit(
    train_generator2,
    steps_per_epoch=steps_per_epoch_training,
    epochs=num_epochs,
    validation_data=validation_generator2,
    validation_steps=steps_per_epoch_validation,
    verbose=1,
)

Epoch 1/5
301/301 [==============================] - 136s 451ms/step - loss: 0.0069 - accuracy: 0.9984
Epoch 2/5
301/301 [==============================] - 136s 452ms/step - loss: 0.0062 - accuracy: 0.9985
Epoch 3/5
301/301 [==============================] - 136s 451ms/step - loss: 0.0056 - accuracy: 0.9988
Epoch 4/5
301/301 [==============================] - 136s 452ms/step - loss: 0.0051 - accuracy: 0.9988
Epoch 5/5
301/301 [==============================] - 136s 451ms/step - loss: 0.0048 - accuracy: 0.9991


# **Evaluate Generator**

In [ ]:
 evaluate_generator = data_generator.flow_from_directory(
    'concrete_data_week4/test',
    target_size=(image_resize, image_resize),
    batch_size=batch_size_training,
    class_mode='categorical',
    shuffle=False)

Found 500 images belonging to 2 classes.


In [ ]:
steps_evaluating = len(evaluate_generator)

## **Restnet50 Evaluate**

In [ ]:
evaluate_restnet50 = model.evaluate(
    evaluate_generator,
    steps=steps_evaluating,
    batch_size=100,
    verbose=1
)

5/5 [==============================] - 2s 311ms/step - loss: 3.0258e-04 - accuracy: 1.0000


## **VGG16 Evaluate**

In [ ]:
evaluate_vgg16 = model2.evaluate(
    evaluate_generator,
    steps=steps_evaluating,
    batch_size=100,
    verbose=1
)

5/5 [==============================] - 2s 405ms/step - loss: 0.0045 - accuracy: 1.0000


## **Evaluate Result**
### **Restnet50**
Using **CPU** time for per epoch running takes about **2 hours**\
Using **GPU** time for per epoch running takes about **144 seconds**\
**Training Accuracy** is **99.85%**\
**Test Accuracy** is **100%** and Test Loss is **3.0258e-04**


### **VGG16**
Using **CPU** time for per epoch running takes about **4 hours**\
Using **GPU** time for per epoch running takes about **136 seconds**\
**Training Accuracy** is **99.91%**\
**Test Accuracy** is **100%** and Test Loss is **0.0045**

Using CPU Restnet50 has shorter time than VGG16, while using GPU Restnet50 has longer time than VGG16. VGG16 Training Accuracy has higher value while VGG16 Loss has bigger value.

### **Summary**

The result between both model for this data is almost the same. But if we compare from the CPU vs GPU performance. In this model for this data VGG16 has better pararel performance. It shows from GPU Performance that has shorter time to fit the model. But if the model run on the CPU (Not focusing on pararel) VGG16 need twice time to fit the model than the Restnet50.


# **Print 5 Predict Result From Test Dataset**

In [ ]:
Test5 = [file for file in os.listdir('concrete_data_week4/test')[:5]]

In [ ]:
 evaluate_generator = data_generator.flow_from_directory(
    'concrete_data_week4/test',
    target_size=(image_resize, image_resize),
    batch_size=batch_size_training,
    class_mode='categorical',

## **Print 5 Predict Result Using Restnet 50 Model**

In [ ]:
predict_restnet50 = model.predict(
    evaluate_generator, verbose=1
)




5/5 [==============================] - 3s 377ms/step


In [ ]:
for x in predict_restnet50[:5]:
    if int(x[0].round()) == 1:
        print('Negative')
    else:
        print('Positive')


Negative
Negative
Negative
Negative
Negative


## **Print 5 Predict Result Using VGG16 Model**

In [ ]:
predict_vgg16 = model2.predict(
    evaluate_generator, verbose=1
)




5/5 [==============================] - 3s 558ms/step


In [ ]:
for x in predict_vgg16[:5]:
    if int(x[0].round()) == 1:
        print('Negative')
    else:
        print('Positive')


Negative
Negative
Negative
Negative
Negative
